core/src/main/java/com/google/cloud/sql/core/ServiceAccountImpersonatingCredentialFactory.java

```diff
@@ -0,0 +1,98 @@
+import com.google.auth.http.HttpCredentialsAdapter;
+import com.google.auth.oauth2.GoogleCredentials;
+import com.google.auth.oauth2.ImpersonatedCredentials;
+import com.google.cloud.sql.CredentialFactory;
+import com.google.common.collect.Lists;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Wraps an existing CredentialFactory, adding service account impersonation for the list of
+ * delegates.
+ *
+ * <p>The "delegates" property is built to work like the cloud-sql-proxy
+ * --impersonate-service-account and gcloud --impersonate-service-account flags: The first element
+ * in the list is the target service account. Intermediate delegated credentials are applied first
+ * from end of the list, working towards the beginning of the list. For example, if <code>
+ * delegates = Arrays.asList( "first@serviceaccount.com", "second@serviceaccount.com",
+ * "third@serviceaccount.com")
+ * </code>
+ *
+ * <p>The connector will start with the GoogleCredentials supplied by `source`. It will use those to
+ * impersonate "third@serviceaccount.com", then with a token for "third@serviceaccount.com",
+ * impersonate "second@serviceaccount.com", and then impersonate "first@serviceaccount.com".
+ * Finally, the connector will attempt to access the services as "first@serviceaccount.com".
+ */
+class ServiceAccountImpersonatingCredentialFactory implements CredentialFactory {
+
+  private final CredentialFactory source;
+  private final List<String> delegates;
+  private final String targetPrincipal;
+
+  ServiceAccountImpersonatingCredentialFactory(CredentialFactory source, List<String> delegates) {
+    if (delegates == null || delegates.isEmpty()) {
+      throw new IllegalArgumentException("delegates must not be empty");
+    }
+    this.source = source;
+
+    // The "delegates" property is built so that delegated credentials are applied first from end
+    // of the list, working towards the beginning of the list. However, the
+    // ImpersonatedCredentials.setDelegates() expects the list to be in the opposite order,
+    // so we have to reverse the list.
+    //
+    // From the ImpersonatedCredentials doc:
+    //
+    //    targetPrincipal – the service account to impersonate
+    //    delegates – the chained list of delegates required to grant the final access_token.
+    //      If set, the sequence of identities must have "Service Account Token Creator"
+    //      capability granted to the preceding identity. For example, if set to
+    //      [serviceAccountB, serviceAccountC], the sourceCredential must have the Token Creator
+    //      role on serviceAccountB. serviceAccountB must have the Token Creator on
+    //      serviceAccountC. Finally, C must have Token Creator on target_principal.
+    //      If unset, sourceCredential must have that role on targetPrincipal.
+    List<String> reversedDelegates = Lists.reverse(delegates);
+    this.targetPrincipal = reversedDelegates.get(reversedDelegates.size() - 1);
```

Comment:
Should this just be `reversedDelegates.get(0)`?

Reply:
No. 
- First, this method receives `delegates` in accordance with the `--impersonate-service-account` contract, so  that the target credential is the first element.
- Next we reverse the list so that it is in order according to the Java ImpersonatedCredentials contract. The target credential is now the last element in `reversedDelegates`. 
- Then set this`this.targetPrincipal` to the last element in `reversedDelegates`
- Finally we set `this.delegates` to the beginning of the list, excluding the last element.

Reply:
Sorry I misspoke. I mean, should this just be `delegates.get(0)`?

Reply:
Well, that would return the same thing. But I think it is more straightforward to only use reversedDelegates.